### PR TITLE
Add video support to Index Exchange Bid Adapter

### PR DESCRIPTION
**Pull request by arvending**
## Type of change

- [ ] Bugfix
- [x] Feature
- [ ] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other

## Description of change
Add video format support to the indexExchangeBidAdapter
Documentation: https://github.com/prebid/prebid.github.io/pull/257

- Change your /etc/hosts to point as.casalemedia.com to 192.139.80.22 (Or create a redirect rule using using a browser plugin to point as.casalemedia.com to sandbox.ht.indexexchange.com)
```
192.139.80.22        as.casalemedia.com
```
- test parameters for validating bids
```
{
  bidder: 'indexExchange',
  params: {
    video: {
      siteID: 188233,
      playerType: "HTML5",
      protocols: ["VAST2", "VAST3"],
      maxduration: 15,
      allowVPAID: true
    }
  }
}
```

## Other information

Documentation at the moment is for video only, will be merged with `indexExchange.md` soon

### CODE REVIEW

File: modules/indexExchangeBidAdapter.js [hunk 1]
```diff
@@ -663,11 +815,304 @@ var IndexExchangeAdapter = function IndexExchangeAdapter() {
...
+            bidmanager.addBidResponse(prebidRequest.placementCode, bid);
+            return;
+          }
+
+          let bid = createBidObj(STATUS.GOOD, prebidRequest);
+          bid.cpm = cpm;
+          bid.width = cygnusRequest.video.w;
+          bid.height = cygnusRequest.video.h;
+          bid.vastUrl = cygnusBid.ext.vasturl;
+          bid.descriptionUrl = cygnusBid.ext.vasturl;
```
Comment (matthewlane) [inline modules/indexExchangeBidAdapter.js]: For video bids, please also set `bid.mediaType = 'video'`

Comment (arvending) [inline modules/indexExchangeBidAdapter.js]: This was updated, thanks

File: modules/indexExchangeBidAdapter.js [hunk 2]
```diff
@@ -425,132 +488,36 @@ var IndexExchangeAdapter = function IndexExchangeAdapter() {
...
-            cygnus_index_args.slots = mergeSlotInto({
-              id: 'T2_' + slotName,
-              width: size.width,
-              height: size.height,
-              siteID: tier3SiteID
-            }, cygnus_index_args.slots);
-          }
+    // Grab the slot level data for cygnus_index_args
+    for (var i = 0; i < bidArr.length; i++) {
+      let bid = bidArr[i];
```
Comment (snapwich) [inline modules/indexExchangeBidAdapter.js]: if you're just iterating over an array for references to each item, you should use `bidArr.forEach(bid =>{})`

File: modules/indexExchangeBidAdapter.js [hunk 3]
```diff
@@ -663,11 +815,305 @@ var IndexExchangeAdapter = function IndexExchangeAdapter() {
...
+    if (!response || !response.seatbid || utils.isEmpty(response.seatbid)) {
+      utils.logInfo('Cygnus returned no bids');
+
+      // signal this response is complete
+      Object.keys(bidRequests)
+        .forEach(bidId => {
+          let prebidRequest = bidRequests[bidId].prebid;
+          let bid = createBidObj(STATUS.NO_BID, prebidRequest);
+          utils.logInfo(JSON.stringify(bid));
+          bidmanager.addBidResponse(prebidRequest.placementCode, bid);
```
Comment (snapwich) [inline modules/indexExchangeBidAdapter.js]: Maybe I'm missing something, but I don't see anywhere where you're clearing out stale bids from `bidRequests`, yet you're iterating through all bids in `bidRequests` here.  Won't you just resubmit a bunch of old bids if multiple auctions are held?

File: modules/indexExchangeBidAdapter.js [hunk 4]
```diff
@@ -663,11 +815,305 @@ var IndexExchangeAdapter = function IndexExchangeAdapter() {
...
+        sizes.push(sizeObj);
+      }
+    }
+
+    return sizes;
+  }
+
   return {
     callBids: _callBids
   };
```
Comment (snapwich) [inline modules/indexExchangeBidAdapter.js]: You need to return the original `this` instance.

You can use
```
return Object.assign(this, {
  callBids: _callBids
});
```
to accomplish this

**Comment by arvending** [comment]
@snapwich All the issues have been resolved.

**Pull request merged**